### PR TITLE
[Notifier] add iqsms bridge

### PR DESCRIPTION
**Pull request by alexandrbarabolia**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/14555
Hi,

I've created integration to notifier to support russian sms operator - [iqsms](https://iqsms.ru)

Can you grab this code and make as symfony/iqsms-notifier?

This PR includes changes in notifier and framework-bundle to support smsapi transport as well as other included in notifier component.

Could someone integrate this into notifier component?

### CODE REVIEW

File: src/Symfony/Component/Notifier/Bridge/Iqsms/CHANGELOG.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+CHANGELOG
+=========
+
+5.2.0
```
Comment (derrabus) [inline src/Symfony/Component/Notifier/Bridge/Iqsms/CHANGELOG.md]: 5.2 is closed for features, that would be 5.3 then.

**Review by derrabus — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Notifier/Bridge/Iqsms/phpunit.xml.dist [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+         bootstrap="vendor/autoload.php"
+         failOnRisky="true"
+         failOnWarning="true"
+>
+    <php>
+        <ini name="error_reporting" value="-1" />
+    </php>
+
+    <testsuites>
+        <testsuite name="Symfony Nexmo Notifier Bridge Test Suite">
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Iqsms/phpunit.xml.dist]:
```suggestion
        <testsuite name="Symfony Iqsms Notifier Bridge Test Suite">
```

File: src/Symfony/Component/Notifier/Bridge/Iqsms/IqsmsTransport.php [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        ]);
+
+        $result = $response->toArray(false);
+        foreach ($result['messages'] as $msg) {
+            if ($msg['status'] === static::STATUS_ACCEPTED) {
+                throw new TransportException(sprintf('Unable to send the SMS: "%s".', $msg['status']), $response);
+            }
+        }
+
+        $success = $response->toArray(false);
```
Comment (jschaedl) [inline src/Symfony/Component/Notifier/Bridge/Iqsms/IqsmsTransport.php]: Do we need this line? Looks the same like line 78.

File: src/Symfony/Component/Notifier/Bridge/Iqsms/README.md [hunk 4]
```diff
@@ -0,0 +1,12 @@
+Iqsms Notifier
+==============
+
+Provides Iqsms integration for Symfony Notifier.
```
Comment (fabpot) [inline src/Symfony/Component/Notifier/Bridge/Iqsms/README.md]: Can you add a note about the DSN format (like done in some other providers)?

**Review by fabpot — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Notifier/Bridge/Iqsms/README.md (hunk 4 above)
Comment (alexandrbarabolia) [reply src/Symfony/Component/Notifier/Bridge/Iqsms/README.md]: Thank for your review. Note already added.

File: src/Symfony/Component/Notifier/Bridge/Iqsms/README.md [hunk 5]
```diff
@@ -3,6 +3,19 @@ Iqsms Notifier
...
+-----------
+
+```
+// .env file
+IQSMS_DSN=iqsms://LOGIN:PASSWORD@default?from=FROM
+```
+
+where:
+ - `LOGIN` is your IQSMS login
+ - `PASSWORD` is your PASSWORD password
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Iqsms/README.md]:
```suggestion
 - `PASSWORD` is your password
```

Comment (alexandrbarabolia) [reply src/Symfony/Component/Notifier/Bridge/Iqsms/README.md]: This is already resolve in https://github.com/symfony/symfony/pull/39096/commits/e667e8fcc0bac0990710a3532ed8343e96d735a2

File: src/Symfony/Component/Notifier/Bridge/Iqsms/LICENSE [hunk 6]
```diff
@@ -0,0 +1,19 @@
+Copyright (c) 2019-2020 Fabien Potencier
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Iqsms/LICENSE]:
```suggestion
Copyright (c) 2020 Fabien Potencier
```

File: src/Symfony/Component/Notifier/Bridge/Iqsms/README.md [hunk 7]
```diff
@@ -0,0 +1,25 @@
+Iqsms Notifier
+==============
+
+Provides Iqsms integration for Symfony Notifier.
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Iqsms/README.md]:
```suggestion
Provides [Iqsms[(https://iqsms.ru) integration for Symfony Notifier.
```

File: src/Symfony/Component/Notifier/Bridge/Iqsms/README.md [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+
+DSN example
+-----------
+
+```
+// .env file
+IQSMS_DSN=iqsms://LOGIN:PASSWORD@default?from=FROM
+```
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Iqsms/README.md]:
```suggestion
```ini
IQSMS_DSN=iqsms://LOGIN:PASSWORD@default?from=FROM
```
```

File: src/Symfony/Component/Notifier/Bridge/Iqsms/README.md [hunk 9]
```diff
@@ -0,0 +1,23 @@
+Iqsms Notifier
+==============
+
+Provides [Iqsms[(https://iqsms.ru) integration for Symfony Notifier.
+
+DSN example
+-----------
+
+```ini
+IQSMS_DSN=iqsms://LOGIN:PASSWORD@default?from=FROM
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Iqsms/README.md:10]:
```suggestion
```ini
IQSMS_DSN=iqsms://LOGIN:PASSWORD@default?from=FROM
```
```

File: src/Symfony/Component/Notifier/Bridge/Iqsms/README.md [hunk 10]
```diff
@@ -0,0 +1,23 @@
+Iqsms Notifier
+==============
+
+Provides [Iqsms[(https://iqsms.ru) integration for Symfony Notifier.
+
+DSN example
+-----------
+
+```ini
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Iqsms/README.md:9]:
```suggestion
```
```

File: src/Symfony/Component/Notifier/Bridge/Iqsms/README.md [hunk 11]
```diff
@@ -0,0 +1,23 @@
+Iqsms Notifier
+==============
+
+Provides [Iqsms[(https://iqsms.ru) integration for Symfony Notifier.
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Iqsms/README.md:4]:
```suggestion
Provides [Iqsms](https://iqsms.ru) integration for Symfony Notifier.
```

File: src/Symfony/Component/Notifier/Bridge/Iqsms/README.md (hunk 10 above)
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Iqsms/README.md]: Sorry, please remove the `ini` and close

Comment (alexandrbarabolia) [reply src/Symfony/Component/Notifier/Bridge/Iqsms/README.md]: ok

File: src/Symfony/Component/Notifier/Bridge/Iqsms/IqsmsTransport.php [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+/**
+ * @author Oleksandr Barabolia <alexandrbarabolya@gmail.com>
+ *
+ * @experimental in 5.3
+ */
+final class IqsmsTransport extends AbstractTransport
+{
+    protected const HOST = 'api.iqsms.ru';
+
+    protected const STATUS_ACCEPTED = 'accepted';
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Iqsms/IqsmsTransport.php]: Can this be private?

File: src/Symfony/Component/Notifier/Bridge/Iqsms/IqsmsTransportFactory.php [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+{
+    /**
+     * @return IqsmsTransport
+     */
+    public function create(Dsn $dsn): TransportInterface
+    {
+        $scheme = $dsn->getScheme();
+        $login = $this->getUser($dsn);
+        $password = $this->getPassword($dsn);
+        $from = $dsn->getOption('from');
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Iqsms/IqsmsTransportFactory.php:40]:
```suggestion
        $from = $dsn->getOption('from');
        
        if (!$from) {
            throw new IncompleteDsnException('Missing from.', $dsn->getOriginalDsn());
        }        
```

File: src/Symfony/Component/Notifier/Bridge/Iqsms/README.md [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+
+DSN example
+-----------
+
+```
+IQSMS_DSN=iqsms://LOGIN:PASSWORD@default?from=FROM
+
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Iqsms/README.md:12]: Please close the code block

File: src/Symfony/Component/Notifier/Bridge/Iqsms/IqsmsTransport.php (hunk 12 above)
Comment (alexandrbarabolia) [reply src/Symfony/Component/Notifier/Bridge/Iqsms/IqsmsTransport.php]: ok

File: src/Symfony/Component/Notifier/Bridge/Iqsms/README.md (hunk 14 above)
Comment (alexandrbarabolia) [reply src/Symfony/Component/Notifier/Bridge/Iqsms/README.md:12]: done

File: src/Symfony/Component/Notifier/Bridge/Iqsms/IqsmsTransport.php [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+
+    public function supports(MessageInterface $message): bool
+    {
+        return $message instanceof SmsMessage;
+    }
+
+    protected function doSend(MessageInterface $message): SentMessage
+    {
+        if (!$message instanceof SmsMessage) {
+            throw new LogicException(sprintf('The "%s" transport only supports instances of "%s" (instance of "%s" given).', __CLASS__, SmsMessage::class, get_debug_type($message)));
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Iqsms/IqsmsTransport.php]: Please use newly introduced `UnsupportedMessageTypeException` here.

Fore reference: https://github.com/symfony/symfony/pull/39365

Thanks

File: src/Symfony/Component/Notifier/Bridge/Iqsms/IqsmsTransportFactory.php [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+        $port = $dsn->getPort();
+
+        if ('iqsms' === $scheme) {
+            return (new IqsmsTransport($login, $password, $from, $this->client, $this->dispatcher))
+                ->setHost($host)
+                ->setPort($port);
+        }
+
+        throw new UnsupportedSchemeException($dsn, 'iqsms', $this->getSupportedSchemes());
+    }
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Iqsms/IqsmsTransportFactory.php]: Please pull up the scheme check like here, to first check if the scheme is supported and afterwards for missing options. Thanks

https://github.com/symfony/symfony/blob/7d846d3e82f4e3dabec7d4b5300f50a1dd9ed7d6/src/Symfony/Component/Notifier/Bridge/Sinch/SinchTransportFactory.php#L27-L47

File: src/Symfony/Component/Notifier/Bridge/Iqsms/IqsmsTransport.php (hunk 15 above)
Comment (alexandrbarabolia) [reply src/Symfony/Component/Notifier/Bridge/Iqsms/IqsmsTransport.php]: оk

File: src/Symfony/Component/Notifier/Bridge/Iqsms/IqsmsTransportFactory.php [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+        if (!$from) {
+            throw new IncompleteDsnException('Missing from.', $dsn->getOriginalDsn());
+        }
+
+        $host = 'default' === $dsn->getHost() ? null : $dsn->getHost();
+        $port = $dsn->getPort();
+
+        return (new IqsmsTransport($login, $password, $from, $this->client, $this->dispatcher))
+            ->setHost($host)
+            ->setPort($port);
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Iqsms/IqsmsTransportFactory.php]:
```suggestion
        return (new IqsmsTransport($login, $password, $from, $this->client, $this->dispatcher))->setHost($host)->setPort($port);
```

**Comment by OskarStark** [comment]
Please add some tests, thank you

**Comment by alexandrbarabolia** [comment]
> Please add some tests, thank you

already added, thank for review

File: src/Symfony/Component/Notifier/Bridge/Iqsms/IqsmsTransport.php [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+                    ],
+                ],
+                'login' => $this->login,
+                'password' => $this->password,
+            ],
+        ]);
+
+        $result = $response->toArray(false);
+        foreach ($result['messages'] as $msg) {
+            if ($msg['status'] !== static::STATUS_ACCEPTED) {
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Iqsms/IqsmsTransport.php]:
```suggestion
            if ('accepted' !== $msg['status']) {
```

File: src/Symfony/Component/Notifier/Bridge/Iqsms/IqsmsTransport.php [hunk 19]
```diff
@@ -0,0 +1,90 @@
...
+ * @author Oleksandr Barabolia <alexandrbarabolya@gmail.com>
+ *
+ * @experimental in 5.3
+ */
+final class IqsmsTransport extends AbstractTransport
+{
+    protected const HOST = 'api.iqsms.ru';
+
+    private const STATUS_ACCEPTED = 'accepted';
+
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Iqsms/IqsmsTransport.php]:
```suggestion
```

File: src/Symfony/Component/Notifier/Bridge/Iqsms/Tests/IqsmsTransportFactoryTest.php [hunk 20]
```diff
@@ -0,0 +1,83 @@
...
+
+final class IqsmsTransportFactoryTest extends TestCase
+{
+    public function testCreateWithDsn()
+    {
+        $factory = $this->createFactory();
+
+        $dsn = 'iqsms://login:password@defaultf?from=some';
+        $transport = $factory->create(Dsn::fromString($dsn));
+        $transport->setHost('host.test');
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Iqsms/Tests/IqsmsTransportFactoryTest.php]:
```suggestion
        $transport = $factory->create(Dsn::fromString($'iqsms://login:password@host.test?from=some'));
```

File: src/Symfony/Component/Notifier/Bridge/Iqsms/Tests/IqsmsTransportFactoryTest.php [hunk 21]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    public function testCreateWithMissingOptionFromThrowsIncompleteDsnException()
+    {
+        $factory = $this->createFactory();
+
+        $this->expectException(IncompleteDsnException::class);
+
+        $dsnIncomplete = 'iqsms://login:password@default';
+        $factory->create(Dsn::fromString($dsnIncomplete));
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Iqsms/Tests/IqsmsTransportFactoryTest.php]:
```suggestion
        $factory->create(Dsn::fromString('iqsms://login:password@default'));
```

File: src/Symfony/Component/Notifier/Bridge/Iqsms/Tests/IqsmsTransportFactoryTest.php [hunk 22]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    public function testSupportsReturnsTrueWithSupportedScheme()
+    {
+        $factory = $this->createFactory();
+
+        $dsn = 'iqsms://login:password@default?from=some';
+        $this->assertTrue($factory->supports(Dsn::fromString($dsn)));
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Iqsms/Tests/IqsmsTransportFactoryTest.php]:
```suggestion
        $this->assertTrue($factory->supports(Dsn::fromString('iqsms://login:password@default?from=some')));
```

File: src/Symfony/Component/Notifier/Bridge/Iqsms/Tests/IqsmsTransportFactoryTest.php [hunk 23]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    public function testSupportsReturnsFalseWithUnsupportedScheme()
+    {
+        $factory = $this->createFactory();
+
+        $dsnUnsupported = 'iiqsms://login:password@default?from=some';
+        $this->assertFalse($factory->supports(Dsn::fromString($dsnUnsupported)));
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Iqsms/Tests/IqsmsTransportFactoryTest.php]:
```suggestion
        $this->assertFalse($factory->supports(Dsn::fromString('somethingElse://login:password@default?from=some')));
```

File: src/Symfony/Component/Notifier/Bridge/Iqsms/Tests/IqsmsTransportFactoryTest.php [hunk 24]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    public function testUnsupportedSchemeThrowsUnsupportedSchemeException()
+    {
+        $factory = $this->createFactory();
+
+        $this->expectException(UnsupportedSchemeException::class);
+
+        $dsnUnsupported = 'iiqsms://login:password@default?from=some';
+        $factory->create(Dsn::fromString($dsnUnsupported));
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Iqsms/Tests/IqsmsTransportFactoryTest.php]:
```suggestion
        $factory->create(Dsn::fromString('somethingElse://login:password@default?from=some'));
```

File: src/Symfony/Component/Notifier/Bridge/Iqsms/Tests/IqsmsTransportFactoryTest.php [hunk 25]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    public function testUnsupportedSchemeThrowsUnsupportedSchemeExceptionEvenIfRequiredOptionIsMissing()
+    {
+        $factory = $this->createFactory();
+
+        $this->expectException(UnsupportedSchemeException::class);
+
+        // unsupported scheme and missing "from" option
+        $factory->create(Dsn::fromString('iiqsms://login:password@default'));
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Iqsms/Tests/IqsmsTransportFactoryTest.php]:
```suggestion
        $factory->create(Dsn::fromString('somethingElse://login:password@default'));
```

File: src/Symfony/Component/Notifier/Bridge/Iqsms/composer.json [hunk 26]
```diff
@@ -0,0 +1,34 @@
...
+        },
+        {
+            "name": "Symfony Community",
+            "homepage": "https://symfony.com/contributors"
+        }
+    ],
+    "require": {
+        "php": ">=7.2.5",
+        "symfony/http-client": "^4.3|^5.0",
+        "symfony/notifier": "^5.2"
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Iqsms/composer.json]:
```suggestion
        "symfony/notifier": "^5.3"
```

**Review by OskarStark — CHANGES_REQUESTED** [review]
Some last changes, afterwards good to go 👍 

Thank you for your patience ❤️

**Comment by OskarStark** [comment]
I added a recipe for the new bridge: https://github.com/symfony/recipes/pull/860

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @alexandrbarabolia.

**Pull request merged**